test/hotspot/jtreg/testlibrary_tests/template_framework/examples/TestTutorial.java

```diff
@@ -0,0 +1,539 @@
+        comp.addJavaSourceCode("p.xyz.InnerTest7", generateWithNames());
+
+        // Compile the source files.
+        comp.compile();
+
+        // Object ret = p.xyz.InnterTest1.main();
+        comp.invoke("p.xyz.InnerTest1", "main", new Object[] {});
+        comp.invoke("p.xyz.InnerTest2", "main", new Object[] {});
+        comp.invoke("p.xyz.InnerTest3", "main", new Object[] {});
+        comp.invoke("p.xyz.InnerTest4", "main", new Object[] {});
+        comp.invoke("p.xyz.InnerTest5", "main", new Object[] {});
+        comp.invoke("p.xyz.InnerTest6", "main", new Object[] {});
+        comp.invoke("p.xyz.InnerTest7", "main", new Object[] {});
+    }
+
+    // This example shows the use of various Tokens.
+    public static String generateWithListOfTokens() {
+        // A Template is essencially a function / lambda that produces a
+        // token body, which is a list of Tokens that are concatenated.
+        var templateClass = Template.make(() -> body(
+            // The "body" method is filled by a sequence of "Tokens".
+            // This can be Strings and multi-line Strings, but also
+            // boxed primitives.
+            """
+            package p.xyz;
+
+            public class InnerTest1 {
+                public static void main() {
+                    System.out.println("Hello World!");
+            """,
+            "int a = ", Integer.valueOf(1), ";\n",
+            "float b = ", Float.valueOf(1.5f), ";\n",
+            // Special Float values are "smartly" formatted!
+            "float nan = ", Float.valueOf(Float.POSITIVE_INFINITY), ";\n",
+            "boolean c = ", Boolean.valueOf(true), ";\n",
+            // Lists of Tokens are also allowed:
+            List.of("int ", "d = 5", ";\n"),
+            // That can be great for streaming / mapping over an existing list:
+            List.of(3, 5, 7, 11).stream().map(i -> "System.out.println(" + i + ");\n").toList(),
+            """
+                    System.out.println(a + " " + b + " " + nan + " " + c + " " + d);
+                }
+            }
+            """
+        ));
+
+        // Render templateClass to String.
+        return templateClass.withArgs().render();
+    }
+
+    // This example shows the use of Templates, with and without arguments.
+    public static String generateWithTemplateArguments() {
+        // A Template with no arguments.
+        var templateHello = Template.make(() -> body(
+            """
+            System.out.println("Hello");
+            """
+        ));
+
+        // A Template with a single Integer argument.
+        var templateCompare = Template.make("arg", (Integer arg) -> body(
+            "System.out.println(", arg, ");\n",  // capture arg via lambda argument
+            "System.out.println(#arg);\n",       // capture arg via hashtag replacement
+            "if (#arg != ", arg, ") { throw new RuntimeException(\"mismatch\"); }\n"
```

Comment:
When should I use the lambda argument and when the hashtag replacement? Maybe add a comment here for some guidance or link to later tutorials where it becomes obvious.

Reply:
Honestly, I don't yet have a clear answer for this. Hmm.
I'm not sure this is the best place to give this guidance.

I guess the difference is to use a separate "token" vs a hashtag replacement.
- token: can paste anything. But it requires you to interrupt the string and add commas. That can be a little clunky. And: you can only do a recursive Template call with the token method.
- hashtag: you need it captured as string, either by a template argument or `let`. Does not allow recursive template calls. But it looks a little nicer cosmetically.

Is this somewhat helpful? Maybe I can put that somewhere later in the tutorial? What do you think?

Reply:
Maybe my guidance would be to prefer hashtag, if need be with a `let`. Especially if it is about inserting something on the same line.
If it is on a new line, then the token method looks nicer often. For example if you stream over a list.
And recursive Template calls just have to be "tokens".

Reply:
![image](https://github.com/user-attachments/assets/51183ba0-1fce-4057-b7d4-bad62cade181)
I added this section here.